x/src/lint/engine/mod.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+
+    pub fn level(&self) -> LintLevel {
+        self.level
+    }
+
+    pub fn message(&self) -> &str {
+        &self.message
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+#[allow(dead_code)]
+#[non_exhaustive]
+pub enum LintLevel {
+    Error,
+    Warning,
+    // TODO: add more levels?
+}
+
+impl fmt::Display for LintLevel {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LintLevel::Error => write!(f, "ERROR"),
+            LintLevel::Warning => write!(f, "WARNING"),
+        }
+    }
+}
+
+/// Message source for lints.
+#[derive(Copy, Clone, Debug)]
+pub struct LintSource<'l> {
+    name: &'static str,
+    kind: LintKind<'l>,
+}
+
+impl<'l> LintSource<'l> {
+    pub(super) fn new(name: &'static str, kind: LintKind<'l>) -> Self {
+        Self { name, kind }
+    }
+
+    pub fn name(&self) -> &'static str {
+        self.name
+    }
+
+    pub fn kind(&self) -> LintKind<'l> {
+        self.kind
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum LintKind<'l> {
+    Project,
+    File(&'l Path),
+    Content(&'l Path),
+}
+
+impl<'l> fmt::Display for LintKind<'l> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            LintKind::Project => write!(f, "project"),
+            LintKind::File(path) => write!(f, "file {}", path.display()),
+            LintKind::Content(path) => write!(f, "content {}", path.display()),
+        }
+    }
+}
+
+/// Type alias for the return type for `run` methods.
+pub type Result<T> = result::Result<T, SystemError>;
+
+/// A system error that happened while running a lint.
+#[derive(Debug)]
+#[non_exhaustive]
+pub enum SystemError {
+    Exec {
+        cmd: &'static str,
+        status: ExitStatus,
+    },
+    Io(io::Error),
+}
+
+impl fmt::Display for SystemError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            SystemError::Exec { cmd, status } => match status.code() {
+                Some(code) => write!(f, "'{}' failed with exit code {}", cmd, code),
+                None => write!(f, "'{}' terminated by signal", cmd),
+            },
+            SystemError::Io(err) => write!(f, "IO error: {}", err),
+        }
+    }
+}
+
+impl error::Error for SystemError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            SystemError::Exec { .. } => None,
+            SystemError::Io(err) => Some(err),
+        }
+    }
+}
+
+impl From<io::Error> for SystemError {
+    fn from(err: io::Error) -> Self {
+        SystemError::Io(err)
+    }
+}
+
+pub mod prelude {
+    pub use super::{
+        content::{ContentContext, ContentLinter},
+        file::FileContext,
+        project::ProjectContext,
+        LintFormatter, LintLevel, LintMessage, LintSource, Linter, Result, RunStatus, SkipReason,
+        SystemError,
+    };
+    pub(super) use super::{LintContext, LintKind};
```

Comment:
seems sort of odd to me to include something in the prelude that you don't actually want public everywhere.

Reply:
Yeah I'll remove it and just do an explicit import, thanks.